### PR TITLE
workqueue: fix leak in queue preventing objects from being GCed

### PR DESCRIPTION
**Pull request by howardjohn**
See https://github.com/grpc/grpc-go/issues/4758 for a real world example
of this leaking 2gb+ of data.

Basically, when we do `q.queue[1:]` we are just repositioning the slice.
The underlying array is still active, which contains the object formerly
known as `q.queue[0]`. Because its referencing this object, it will not
be GCed. The only thing that will trigger it to free is eventually when
we add enough to the queue that we allocate a whole new array.

Instead, we should explicitly clear out the old space when we remove it
from the queue. This ensures the object can be GCed, assuming the users'
application doesn't reference it anymore.

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
See https://github.com/grpc/grpc-go/issues/4758 for a real world example
of this leaking 2gb+ of data.

Basically, when we do `q.queue[1:]` we are just repositioning the slice.
The underlying array is still active, which contains the object formerly
known as `q.queue[0]`. Because its referencing this object, it will not
be GCed. The only thing that will trigger it to free is eventually when
we add enough to the queue that we allocate a whole new array.

Instead, we should explicitly clear out the old space when we remove it
from the queue. This ensures the object can be GCed, assuming the users'
application doesn't reference it anymore.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
same fix as https://github.com/kubernetes/kubernetes/pull/104991?

File: staging/src/k8s.io/client-go/util/workqueue/queue_test.go [hunk 1]
```diff
@@ -321,3 +323,40 @@ func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
 
 	finishedWG.Wait()
 }
+
+type bigObject struct {
+	data []byte
+}
+
+// Created outside TestKubeLeak so it won't be GCed.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/workqueue/queue_test.go:331]: should be able to reference from inside the registered Cleanup closure to keep it from going out of scope, without making this a stateful package var, right?

File: staging/src/k8s.io/client-go/util/workqueue/queue_test.go [hunk 2]
```diff
@@ -321,3 +323,40 @@ func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
...
+// mustGarbageCollect asserts than an object was garbage collected by the end of the test.
+// The input must be a pointer to an object.
+func mustGarbageCollect(t *testing.T, i interface{}) {
+	t.Helper()
+	collected := false
+	runtime.SetFinalizer(i, func(x interface{}) {
+		collected = true
+	})
+	t.Cleanup(func() {
+		if err := wait.PollImmediate(time.Millisecond*100, time.Second*5, func() (done bool, err error) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/workqueue/queue_test.go:353]: suggest waiting longer than 5 seconds (`wait.ForeverTestTimeout` comes to mind) to avoid flakes

File: staging/src/k8s.io/client-go/util/workqueue/queue_test.go [hunk 3]
```diff
@@ -321,3 +323,40 @@ func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
...
+
+			// Trigger GC explicitly, otherwise we may need to wait a long time for it to run
+			runtime.GC()
+			return collected, nil
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/workqueue/queue_test.go:357]: can referencing `collected` here and writing it in the finalizer cause a data race?

**Comment by hzxuzhonghu** [comment]
@liggitt I was inspired by @howardjohn 's fix in istio

**Comment by howardjohn** [comment]
@liggitt thanks, addressed the comments. The fix got merge without tests in another PR so I guess this just adds tests now.

File: staging/src/k8s.io/client-go/util/workqueue/queue_test.go [hunk 4]
```diff
@@ -321,3 +324,44 @@ func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
 
 	finishedWG.Wait()
 }
+
+
+
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/workqueue/queue_test.go:329]: drop the "Created outside" comment and run hack/update-gofmt.sh to clean up formatting, then lgtm

File: staging/src/k8s.io/client-go/util/workqueue/queue_test.go [hunk 5]
```diff
@@ -321,3 +324,40 @@ func TestQueueDrainageUsingShutDownWithDrainWithDirtyItem(t *testing.T) {
 
 	finishedWG.Wait()
 }
+
+// TestGarbageCollection ensures that objects that are added then removed from the queue are
+// able to be garbage collected.
+func TestGarbageCollection(t *testing.T) {
+	type bigObject struct {
+		data []byte
```
Comment (pacoxu) [inline staging/src/k8s.io/client-go/util/workqueue/queue_test.go:332]: staticcheck failed
> Errors from staticcheck:
> vendor/k8s.io/client-go/util/workqueue/queue_test.go:332:3: field data is unused (U1000)

/hold 
Feel free to remove the hold once CI is green.
